File: src/main/java/aoc2019/Day01.java
package aoc2019;

import static java.lang.Math.max;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.function.IntUnaryOperator;
import java.util.stream.IntStream;

/*
 * https://adventofcode.com/2019/day/1
 */
public class Day01 {

    public static void main(String[] args) throws Exception {
        new Day01().part1();
        new Day01().part2();
    }
    
    void part1() throws Exception {      
    	var input = Util.intStreamOf("input01.txt");
    	
        IntUnaryOperator fuel = mass -> mass/3 - 2;
        
		var result = input.map(fuel).sum();
        System.out.println(result);
    }

    void part2() throws Exception {
    	var input = Util.intStreamOf("input01.txt");
    	
    	IntUnaryOperator fuel = mass -> max(0, mass/3 - 2);
    	IntUnaryOperator totalFuel = mass -> IntStream.iterate(mass, x -> x>0, fuel).skip(1).sum();
    	
    	assertEquals(50346, totalFuel.applyAsInt(100756));
    	
        var result = input.map(totalFuel).sum();
        System.out.println(result);
    }

}

File: src/main/csharp/aoc2019/Day01.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using aoc;

namespace aoc2019
{
    class Day01 : AocDay
    {

        int[] Data = ReadInts("input01.txt");

        public void main()
        {
            Console.WriteLine("=== AOC 2019 day 01");
            var t0 = DateTime.Now;

            Console.WriteLine("--- part 1");
            Part1();

            Console.WriteLine("--- part 2");
            Part2();

            Console.WriteLine("=== end ({0} ms)", (DateTime.Now - t0).Milliseconds);
        }

        private void Part2()
        {
            int value = Data.SelectMany(RecFuel).Sum();
            Console.WriteLine(value);
        }

        IEnumerable<int> RecFuel(int mass)
        {
            int f = 0;
            while ((f = fuel(mass)) > 0)
            {
                yield return f;
                mass = f;
            }
        }

        void Part1()
        {
            int value = Data.Select(fuel).Sum();
            Console.WriteLine(value);
        }

        int fuel(int mass)
        {
            return (mass / 3) - 2;
        }
    }
}
